### PR TITLE
Performance fixes to TTS

### PR DESCRIPTION
**Pull request by Carbonhell**
:cl: 
refactor: TTS performance has been improved.
/:cl:

check the code carefully, it seemed to work but i did slight edits after testing.

### CODE REVIEW

File: hippiestation/code/controllers/subsystem/tts.dm [hunk 1]
```diff
@@ -4,129 +4,96 @@
...
+	var/audio_length = text2num(file2text(T.filename + ".meta"))
+	audio_length = audio_length / 100
+	if (!audio_length)
+		audio_length = length(T.text)
 
 	if (T.is_global)
-		for (var/mob/M in GLOB.player_list)
-			if (!M.client)
-				continue
+		for (var/mob/M in SSmobs.clients_by_zlevel[origin.z])
```
Comment (JohnGinnane) [inline hippiestation/code/controllers/subsystem/tts.dm]: why have z level check here, global tts should play to everyone?

File: hippiestation/code/controllers/subsystem/tts.dm [hunk 2]
```diff
@@ -4,129 +4,96 @@
...
-		T.life = world.time + audio_length
 		T.owner.tts_cooldown = world.time + audio_length
 
 		addtimer(CALLBACK(T.owner.mob, /mob/living.proc/update_tts_hud), audio_length)
+		addtimer(CALLBACK(src, .proc/delete_files, T), audio_length)
 
-		for (var/mob/P in listeners)
-			if (!P.client)
-				continue
+		for (var/M in listeners)
```
Comment (JohnGinnane) [inline hippiestation/code/controllers/subsystem/tts.dm:85]: `var/mob/P` if you're just gonna convert to mob anyway

Comment (Carbonhell) [reply hippiestation/code/controllers/subsystem/tts.dm:85]: No, this is actually better. By doing for(var/type/T in list), byond grabs each item and istypes is. It is not needed if we know the list only has one type. It is a performance upgrade aye

File: hippiestation/code/controllers/subsystem/tts.dm (hunk 1 above)
Comment (Carbonhell) [reply hippiestation/code/controllers/subsystem/tts.dm]: Shit you're right

File: hippiestation/code/controllers/subsystem/tts.dm (hunk 2 above)
Comment (JohnGinnane) [reply hippiestation/code/controllers/subsystem/tts.dm:85]: alrighty

File: hippiestation/code/controllers/subsystem/tts.dm [hunk 3]
```diff
@@ -4,129 +4,96 @@
...
 		return
 	fdel(T.filename + ".ogg")
 	fdel(T.filename + ".meta")
+	STOP_PROCESSING(src, T)
+	ckeys_playing -= T.ckey
+	qdel(T) // not needed anymore, it can die
 
-/datum/controller/subsystem/tts/proc/play_tts(datum/tts/T)
+/datum/controller/subsystem/processing/tts/proc/play_tts(datum/tts/T)
+	T.status = STATUS_PLAYING
```
Comment (JohnGinnane) [inline hippiestation/code/controllers/subsystem/tts.dm:45]: might want to `if (!T) return` just to be safe but I wonder if all these checks have any noticeable effect on performance

File: hippiestation/code/controllers/subsystem/tts.dm [hunk 4]
```diff
@@ -4,129 +4,96 @@
...
 	if (!T.owner.mob)
 		message_admins("TTS request has no mob")
+		delete_files(T)
 		return
 
 	var/next_channel = open_sound_channel()
+	var/turf/origin = get_turf(T.owner.mob)
 
-	T.status = STATUS_PLAYING
+	if(!origin)
```
Comment (JohnGinnane) [inline hippiestation/code/controllers/subsystem/tts.dm]: Why take this out of the `else`? It was in there so that it wouldn't be redundant for global TTS

**Comment by JohnGinnane** [comment]
So let me get this straight:

Rather than using a subsystem you're converting it to a process?

If someone says something it calls `tts.say` which sets that datum to be processed

The processing subsystem calls `process()` on each TTS datum and this will mean I don't manage the group of datums, I just manage one at a time

**Comment by Carbonhell** [comment]
What do you mean? You already managed them one at a time, the only difference is that the subsystem is a child of processing now, which has some performance gains and acts in the same way, you just kinda recreated it

**Comment by JohnGinnane** [comment]
Yeah but I mean I would have a list of datums that I looped through in `fire()` whereas now you just add them to the processing and then only handle when one is checked via `process()`

**Comment by JohnGinnane** [comment]
so yeah

**Review by JohnGinnane — APPROVED** [review]
hit it

**Comment by JohnGinnane** [comment]
Fixes https://github.com/HippieStation/HippieStation/projects/9#card-12652579

**Pull request merged**